### PR TITLE
Add public link logic

### PR DESCRIPTION
**Pull request by artonge**
:heavy_check_mark: To review

### CODE REVIEW

File: lib/Album/AlbumMapper.php [hunk 1]
```diff
@@ -404,7 +414,7 @@ public function getSharedAlbumsForCollaboratorWithFiles(string $collaboratorId,
 			if ($row['fileid']) {
 				$mimeId = $row['mimetype'];
 				$mimeType = $this->mimeTypeLoader->getMimetypeById($mimeId);
-				$filesByAlbum[$albumId][] = new AlbumFile((int)$row['fileid'], $row['album_name'].' ('.$row['album_user'].')', $mimeType, (int)$row['size'], (int)$row['mtime'], $row['etag'], (int)$row['added'], $row['owner']);
+				$filesByAlbum[$albumId][] = new AlbumFile((int)$row['fileid'], $row['file_name'], $mimeType, (int)$row['size'], (int)$row['mtime'], $row['etag'], (int)$row['added'], $row['owner']);
```
Comment (artonge) [inline lib/Album/AlbumMapper.php:417]: Restore correct behavior. Previous change was intended for `AlbumInfo` on line 420

File: lib/Sabre/Album/PublicAlbumRoot.php [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+	}
+
+	/**
+	 * @return void
+	 */
+	public function setName($name) {
+		throw new Forbidden('Not allowed to rename a public album');
+	}
+
+	// TODO: uncomment else it is a security hole.
```
Comment (artonge) [inline lib/Sabre/Album/PublicAlbumRoot.php:45]: @PVince81 please advise, linting is not happy when uncommented

File: lib/Controller/PublicAlbumController.php [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+		$this->userConfig = $userConfig;
+		$this->initialState = $initialState;
+	}
+
+	/**
+	 * Your normal controller function. The following annotation will allow guests
+	 * to open the page as well
+	 *
+	 * @PublicPage
+	 * @NoAdminRequired
```
Comment (CarlSchwan) [inline lib/Controller/PublicAlbumController.php:68]: Don't think this is needed in a PublicPage

File: lib/Controller/PublicAlbumController.php [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+	 * @NoCSRFRequired
+	 *
+	 * @return TemplateResponse
```
Comment (CarlSchwan) [inline lib/Controller/PublicAlbumController.php:71]:
```suggestion
```

File: lib/Service/UserConfigService.php [hunk 5]
```diff
@@ -52,16 +52,20 @@ public function __construct(
 	}
 
 	public function getUserConfig(string $key) {
+		$user = $this->userSession->getUser();
+		return $this->getConfigForUser($user->getUid(), $key);
+	}
+
+	public function getConfigForUser(string $userId, string $key) {
```
Comment (CarlSchwan) [inline lib/Service/UserConfigService.php:59]:
```suggestion
	public function getConfigForUser(string $userId, string $key): string {
```

File: lib/Sabre/Album/PublicAlbumsHome.php [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+		throw new Forbidden('Not allowed to create folders in this folder');
+	}
+
+	public function getChild($name) {
+		$basicAuth = $this->server->httpRequest->getHeader('Authorization');
+		[, $base64Token] = explode('Basic ', $basicAuth);
+		$token = \base64_decode($base64Token);
+		$albums = $this->albumMapper->getSharedAlbumsForCollaboratorWithFiles($token, AlbumMapper::TYPE_LINK);
+
+		array_filter($albums, fn ($album) => $album->getAlbum()->getUserId() === $this->user->getUid());
```
Comment (artonge) [inline lib/Sabre/Album/PublicAlbumsHome.php:73]: Receive assignment of filter

File: lib/Sabre/Album/AlbumRoot.php [hunk 7]
```diff
@@ -212,12 +212,21 @@ public function getCover() {
...
 			fn (array $collaborator) => [ 'nc:collaborator' => $collaborator ],
 			$this->albumMapper->getCollaborators($this->album->getAlbum()->getId()),
 		);
 	}
+
+	/**
+	 * @param array{'id': string, 'type': int} $collaborators
+	 * @return array{'id': string, 'label': string, 'type': int}
+	 */
+	public function setCollaborators($collaborators) {
```
Comment (artonge) [inline lib/Sabre/Album/AlbumRoot.php:228]: return type

Comment (artonge) [reply lib/Sabre/Album/AlbumRoot.php:228]: : array

File: src/components/Albums/CollaboratorsSelectionForm.vue [hunk 8]
```diff
@@ -183,15 +184,24 @@ export default {
...
 			currentSearchResults: [],
 			loadingCollaborators: false,
 			randomId: Math.random().toString().substring(2, 10),
 			publicLinkCopied: false,
 			config: {
 				minSearchStringLength: parseInt(OC.config['sharing.minSearchStringLength'], 10) || 0,
 			},
+			/** @type {Collaborator} */
+			publicLink: {
+				id: (Math.random().toString(36).substring(2) + Math.random().toString(36).substring(2)).substring(0, 15),
```
Comment (artonge) [inline src/components/Albums/CollaboratorsSelectionForm.vue:201]: generate token in the backend

File: lib/Sabre/Album/SharedAlbumRoot.php [hunk 9]
```diff
@@ -59,4 +59,9 @@ protected function addFile(int $sourceId, string $ownerUID): bool {
 		$this->albumMapper->addFile($this->album->getAlbum()->getId(), $sourceId, $ownerUID);
 		return true;
 	}
+
+	// Do not reveal collaborators for shared albums.
+	public function getCollaborators() {
```
Comment (artonge) [inline lib/Sabre/Album/SharedAlbumRoot.php:64]: typing

File: lib/Sabre/Album/PublicAlbumRoot.php [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+
+	/**
+	 * We cannot create files in an Album
+	 * We add the file to the default Photos folder and then link it there.
+	 *
+	 * @param [type] $name
+	 * @param [type] $data
+	 * @return void
+	 */
+	public function createFile($name, $data = null) {
```
Comment (artonge) [inline lib/Sabre/Album/PublicAlbumRoot.php:58]: remove as we don't do upload for public albums

File: src/PhotosPublic.vue [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+	async beforeMount() {
+		if ('serviceWorker' in navigator) {
+			// Use the window load event to keep the page load performant
+			window.addEventListener('load', () => {
+				navigator.serviceWorker.register(generateUrl('/apps/photos/service-worker.js', {}, {
+					noRewrite: true,
+				}), {
+					scope: '/',
+				}).then(registration => {
+					console.debug('SW registered: ', registration)
```
Comment (CarlSchwan) [inline src/PhotosPublic.vue:71]: use `import { getLoggerBuilder } from '@nextcloud/logger'`

File: lib/Sabre/Album/PublicAlbumRoot.php [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+			// https://github.com/nextcloud/server/blob/af29b978078ffd9169a9bd9146feccbb7974c900/apps/dav/lib/Connector/Sabre/FilesPlugin.php#L564-L585
+			\header('OC-FileId: ' . $node->getId());
+			return '"' . $node->getEtag() . '"';
+		} catch (\Exception $e) {
+			throw new Forbidden('Could not create file');
+		}
+	}
+
+
+	protected function addFile(int $sourceId, string $ownerUID): bool {
```
Comment (artonge) [inline lib/Sabre/Album/PublicAlbumRoot.php:82]: Remose, same as above

File: lib/Sabre/Album/PublicAlbumRoot.php [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+		if (in_array($sourceId, $this->album->getFileIds())) {
+			throw new Conflict("File $sourceId is already in the folder");
+		}
+
+		$this->albumMapper->addFile($this->album->getAlbum()->getId(), $sourceId, $ownerUID);
+		return true;
+	}
+
+	// Do not reveal collaborators for public albums.
+	public function getCollaborators() {
```
Comment (artonge) [inline lib/Sabre/Album/PublicAlbumRoot.php:92]: typing

File: src/PhotosPublic.vue [hunk 14]
```diff
@@ -0,0 +1,102 @@
+<!--
+ - @copyright Copyright (c) 2019 John Molakvoæ <skjnldsv@protonmail.com>
```
Comment (artonge) [inline src/PhotosPublic.vue:2]: impersonation

File: lib/Controller/PublicAlbumController.php [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+		$this->userConfig = $userConfig;
+		$this->initialState = $initialState;
+	}
+
+	/**
+	 * @PublicPage
+	 * @NoCSRFRequired
+	 */
+	public function get(): TemplateResponse {
+		$this->eventDispatcher->dispatch(LoadSidebar::class, new LoadSidebar());
```
Comment (skjnldsv) [inline lib/Controller/PublicAlbumController.php:66]: Is there a sidebar in public mode?

File: lib/Controller/PublicAlbumController.php [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+		$this->eventDispatcher->dispatch(LoadViewer::class, new LoadViewer());
+
+		$this->initialState->provideInitialState('image-mimes', Application::IMAGE_MIMES);
+		$this->initialState->provideInitialState('video-mimes', Application::VIDEO_MIMES);
+		$this->initialState->provideInitialState('maps', $this->appManager->isEnabledForUser('maps') === true);
+		$this->initialState->provideInitialState('recognize', $this->appManager->isEnabledForUser('recognize') === true);
+		$this->initialState->provideInitialState('systemtags', $this->appManager->isEnabledForUser('systemtags') === true);
```
Comment (skjnldsv) [inline lib/Controller/PublicAlbumController.php:73]: Do they even work in public mode?

File: lib/Sabre/Album/AlbumRoot.php [hunk 17]
```diff
@@ -212,12 +212,21 @@ public function getCover() {
 	}
 
 	/**
-	 * @return array
+	 * @return array{array{'nc:collaborator': array{'id': string, 'label': string, 'type': int}}}
```
Comment (skjnldsv) [inline lib/Sabre/Album/AlbumRoot.php:215]: Maybe we would gain by creating a Collaborator object instead?

File: lib/Sabre/PhotosHome.php [hunk 18]
```diff
@@ -89,24 +90,31 @@ public function createDirectory($name) {
 	}
 
 	public function getChild($name) {
-		if ($name === 'albums') {
-			return new AlbumsHome($this->principalInfo, $this->albumMapper, $this->user, $this->rootFolder, $this->userConfigService);
-		} elseif ($name === 'sharedalbums') {
-			return new SharedAlbumsHome($this->principalInfo, $this->albumMapper, $this->user, $this->rootFolder, $this->groupManager, $this->userConfigService);
+		switch ($name) {
+			case 'albums':
```
Comment (skjnldsv) [inline lib/Sabre/PhotosHome.php:94]: enum?

File: src/Photos.vue [hunk 19]
```diff
@@ -172,14 +173,14 @@ export default {
 				}), {
 					scope: '/',
 				}).then(registration => {
-					console.debug('SW registered: ', registration)
+					logger.debug('SW registered: ', registration)
```
Comment (skjnldsv) [inline src/Photos.vue:176]:
```suggestion
					logger.debug('SW registered: ', { registration })
```

File: src/Photos.vue [hunk 20]
```diff
@@ -172,14 +173,14 @@ export default {
...
 				}).catch(registrationError => {
-					console.error('SW registration failed: ', registrationError)
+					logger.error('SW registration failed: ', registrationError)
```
Comment (skjnldsv) [inline src/Photos.vue:178]:
```suggestion
					logger.error('SW registration failed: ', { registrationError })
```

File: src/components/Albums/CollaboratorsSelectionForm.vue [hunk 21]
```diff
@@ -275,17 +311,89 @@ export default {
...
+				logger.error('[PublicAlbumContent] Error updating album', error)
+				showError(this.t('photos', 'Failed to update album.'))
+			} finally {
+				this.loadingAlbum = false
+			}
 		},
 
 		async copyPublicLink() {
-			await navigator.clipboard.writeText(this.publicLink)
+			await navigator.clipboard.writeText(`${window.location.protocol}//${window.location.host}${generateUrl(`apps/photos/public/${getCurrentUser().uid}/${this.publicLink.id}`)}`)
```
Comment (skjnldsv) [inline src/components/Albums/CollaboratorsSelectionForm.vue:396]: Isn't it what `generateAbsoluteURL` is for?

File: src/mixins/FetchAlbumsMixin.js [hunk 22]
```diff
@@ -20,16 +20,12 @@
...
 
-import moment from '@nextcloud/moment'
-import { showError } from '@nextcloud/dialogs'
 import { getCurrentUser } from '@nextcloud/auth'
 
-import client from '../services/DavClient.js'
-import logger from '../services/logger.js'
-import { genFileInfo } from '../utils/fileUtils.js'
 import AbortControllerMixin from './AbortControllerMixin.js'
+import { fetchAlbums } from '../services/Albums.js'
```
Comment (skjnldsv) [inline src/mixins/FetchAlbumsMixin.js:28]: :tada: :tada: :tada:

File: src/views/AlbumContent.vue [hunk 23]
```diff
@@ -361,9 +363,7 @@ export default {
 					this.errorFetchingFiles = error
 				}
 
-				// cancelled request, moving on...
-				logger.error('Error fetching album files')
-				console.error(error)
+				logger.error('[AlbumContent] Error fetching album files', error)
```
Comment (skjnldsv) [inline src/views/AlbumContent.vue:366]:
```suggestion
				logger.error('[AlbumContent] Error fetching album files', { error })
```

File: src/views/PublicAlbumContent.vue [hunk 24]
```diff
@@ -0,0 +1,315 @@
...
+				key="navigation"
+				slot="header"
+				slot-scope="{selectedFileIds}"
+				:loading="loadingAlbum || loadingFiles"
+				:params="{ userId, token }"
+				path="/"
+				:root-title="albumOriginalName"
+				:title="albumOriginalName"
+				@refresh="fetchAlbumContent">
+				<!-- TODO: enable upload on public albums -->
```
Comment (skjnldsv) [inline src/views/PublicAlbumContent.vue:41]: Really? :see_no_evil:

File: lib/Service/UserConfigService.php [hunk 25]
```diff
@@ -51,17 +51,21 @@ public function __construct(
...
+	}
+
+	public function getConfigForUser(string $userId, string $key): string {
 		if (!in_array($key, array_keys(self::DEFAULT_CONFIGS))) {
 			throw new Exception('Unknown user config key');
 		}
-		$user = $this->userSession->getUser();
 		$default = self::DEFAULT_CONFIGS[$key];
-		$value = $this->config->getUserValue($user->getUid(), Application::APP_ID, $key, $default);
+		$value = $this->config->getUserValue($userId, Application::APP_ID, $key, $default);
```
Comment (skjnldsv) [inline lib/Service/UserConfigService.php:64]: Revert and check if user === null, if so, return default values

File: lib/Sabre/Album/AlbumPhoto.php [hunk 26]
```diff
@@ -152,9 +152,11 @@ public function setFavoriteState($favoriteState): bool {
 
 		switch ($favoriteState) {
 			case "0":
-				return $tagger->removeFromFavorites($this->albumFile->getFileId());
+				$tagger->removeFromFavorites($this->albumFile->getFileId());
+				return "0";
 			case "1":
-				return $tagger->addToFavorites($this->albumFile->getFileId());
+				$tagger->addToFavorites($this->albumFile->getFileId());
+				return "1";
```
Comment (skjnldsv) [inline lib/Sabre/Album/AlbumPhoto.php:159]: Revert?

File: lib/Sabre/PhotosHome.php (hunk 18 above)
Comment (skjnldsv) [reply lib/Sabre/PhotosHome.php:94]: Use const in each XxxHome

File: lib/Service/UserConfigService.php [hunk 27]
```diff
@@ -51,17 +51,21 @@ public function __construct(
...
 
 		// If the config is a path, make sure it exists
 		if (str_starts_with($default, '/')) {
-			$userFolder = $this->rootFolder->getUserFolder($user->getUID());
+			$userFolder = $this->rootFolder->getUserFolder($userId);
 			// If the folder does not exists, create it
 			if (!$userFolder->nodeExists($value)) {
 				$userFolder->newFolder($value);
```
Comment (skjnldsv) [inline lib/Service/UserConfigService.php:71]: Move Folder creatin logic to a protected in a protected `AlbumRoot`.
So we can call it straight from `AlbumRoot->createFile` as well from `PublicAlbumRoot` and  `SharedAlbumRoot`

File: src/router/index.js [hunk 28]
```diff
@@ -118,6 +119,15 @@ const router = new Router({
...
 			}),
 		},
+		{
+			path: '/public/:userId/:token',
+			component: PublicAlbumContent,
+			name: 'publicAlbums',
+			props: route => ({
+				userId: route.params.userId,
+				token: route.params.token,
+			}),
```
Comment (skjnldsv) [inline src/router/index.js:129]: We should obfuscate the userID

File: src/store/collectionStoreFactory.js [hunk 29]
```diff
@@ -0,0 +1,214 @@
...
+		 * @param {string[]} data.fileIdsToAdd list of files
+		 */
+		[`addFilesTo${capitalizedCollectionName}`](state, { collectionId, fileIdsToAdd }) {
+			const collectionFiles = state[`${collectionName}sFiles`][collectionId] || []
+			state[`${collectionName}sFiles`] = {
+				...state[`${collectionName}sFiles`],
+				[collectionId]: [
+					...collectionFiles,
+					...fileIdsToAdd.filter(fileId => !collectionFiles.includes(fileId)), // Filter to prevent duplicate fileId.
+				],
```
Comment (skjnldsv) [inline src/store/collectionStoreFactory.js:83]: Use new Set ?

File: src/views/SharedAlbumContent.vue [hunk 30]
```diff
@@ -263,7 +263,7 @@ export default {
 				}
 
 				// cancelled request, moving on...
-				logger.error('Error fetching shared album files', error)
+				logger.error('[SharedAlbumContent] Error fetching album files', error)
```
Comment (skjnldsv) [inline src/views/SharedAlbumContent.vue:266]:
```suggestion
				logger.error('[SharedAlbumContent] Error fetching album files', { error })
```

**Comment by artonge** [comment]
Important changes I made after realizing that I didn't test without being logged in…

- Intercept webdav request in the backend to authenticate public access with a token
https://github.com/nextcloud/photos/pull/1278/files#diff-213bbab031403681106d94580acb181d9cde8a708293283a9e620b542c390d8c

- For that, I needed to add type `authentication` in `info.xml` to intercept them early enough
https://github.com/nextcloud/photos/pull/1278/files#diff-ab97a56738913a50d2334c4d3f10991748d3e29d011f5a674734b63d33d5ce9eR15

File: lib/Controller/PreviewController.php [hunk 31]
```diff
@@ -71,45 +71,62 @@ public function __construct(
 	/**
 	 * @NoAdminRequired
 	 * @NoCSRFRequired
+	 * @PublicPage
```
[the remark this answers is not in the document]
Comment (skjnldsv) [reply lib/Controller/PreviewController.php:74]: Ah right, for public pages, we need to serve it, ok.
I would prefer to split the two controlelrs for easier maintenance and security :)

https://github.com/nextcloud/server/blob/2397ea72196de42a79e2a5873dcbb07db698f348/apps/files_sharing/lib/Controller/PublicPreviewController.php#L131-L139

```
	 * @PublicPage
	 * @NoCSRFRequired
	 * @NoSameSiteCookieRequired
```

**Review by skjnldsv — APPROVED** [review]

**Pull request merged**